File: lw7/2/Vasilyev_Andrey/lab7/SimpleMergeSorter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace lab7
{
    public class SimpleMergeSorter: Sorter
    {
        protected List<int> helper;

        public SimpleMergeSorter(List<int> data)
            : base(data)
        {
            helper = new List<int>(new int[Data.Count + 1]);
        }

        public override void Sort()
        {
            MergeSort(0, Data.Count - 1);
        }

        public override string ToString()
        {
            return "Simple mergeSorter";
        }

        protected void MergeSort(int low, int high)
        {
            if (low < high)
            {
                int middle = low + (high - low) / 2;
                MergeSort(low, middle);
                MergeSort(middle + 1, high);
                Merge(low, middle, high);
            }
        }

        protected void Merge(int low, int middle, int high)
        {
            for (int m = low; m <= high; m++)
            {
                helper[m] = Data[m];
            }

            int i = low;
            int j = middle + 1;
            int k = low;

            while (i <= middle && j <= high)
            {
                if (helper[i] <= helper[j])
                {
                    Data[k] = helper[i];
                    i++;
                }
                else
                {
                    Data[k] = helper[j];
                    j++;
                }
                k++;
            }

            while (i <= middle)
            {
                Data[k] = helper[i];
                k++;
                i++;
            }
        }
    }
}

File: lw6/2_Vasilyev_Andrey/SimpleMergeSorter.cpp
#include "SimpleMergeSorter.h"

CSimpleMergeSorter::CSimpleMergeSorter(const std::vector<int> & data)
	:CSorter(data)
{

}

std::string CSimpleMergeSorter::ToString()
{
	return "Simple mergeSorter";
}


void CSimpleMergeSorter::Sort()
{
	helper = std::vector<int>(m_data.size());
	MergeSort(0, m_data.size() - 1);
}

void CSimpleMergeSorter::MergeSort(int low, int high)
{
	if (low < high)
	{
		int middle = low + (high - low) / 2;
		MergeSort(low, middle);
		MergeSort(middle + 1, high);
		Merge(low, middle, high);
	}
}

void CSimpleMergeSorter::Merge(int low, int middle, int high) 
{
	for (int i = low; i <= high; i++) 
	{
		helper[i] = m_data[i];
	}

	int i = low;
	int j = middle + 1;
	int k = low;

	while (i <= middle && j <= high)
	{
		if (helper[i] <= helper[j]) 
		{
				m_data[k] = helper[i];
				i++;
		}
		else 
		{
			m_data[k] = helper[j];
			j++;
		}
		k++;
	}

	while (i <= middle) 
	{
		m_data[k] = helper[i];
		k++;
		i++;
	}
}